modal/_utils/async_utils.py

```diff
@@ -463,3 +464,83 @@ def run_generator_sync(
         except BaseException as err:
             exc = err
     loop.close()
+
+
+@asynccontextmanager
+async def aclosing(gen: AsyncGenerator[T, None]) -> AsyncGenerator[AsyncGenerator[T, None], None]:
+    try:
+        yield gen
+    finally:
+        await gen.aclose()
+
+
+async def async_map(
+    input: AsyncIterator[T], async_mapper_func: Callable[[T], Awaitable[V]], concurrency: int
+) -> AsyncIterator[V]:
+    input_queue: asyncio.Queue[T] = asyncio.Queue(maxsize=concurrency)
+    results_queue: asyncio.Queue[V] = asyncio.Queue()
+    end_sentinel = object(T)
+
+    async def producer():
+        async for item in input:
+            await input_queue.put(item)
+        for _ in range(concurrency):
+            await input_queue.put(end_sentinel)
+
+    async def worker():
+        while True:
+            item = await input_queue.get()
+            if item is end_sentinel:
+                break
+            result = await async_mapper_func(item)
+            await results_queue.put(result)
+
+    producer_task = asyncio.create_task(producer())
+    worker_tasks = [asyncio.create_task(worker()) for _ in range(concurrency)]
+
+    try:
+        while True:
+            if results_queue.empty():
+                if producer_task.done() and all(t.done() for t in worker_tasks):
+                    break
+                await asyncio.sleep(0.01)  # Small delay to avoid busy waiting
```

Comment:
It's still a pretty busy loop... I think it would be better to `asyncio.wait()` to wait for any of the underlying tasks to complete than to use a timed sleep